File: python_base_04/core/modules/recall_game/utils/predefined_hands_loader.py
"""
Predefined Hands Loader for Testing

This module provides functionality to load predefined hands configuration
for testing purposes in the Recall game.
"""

import os
import yaml
from typing import Dict, List, Any, Optional
from utils.logging_utils import custom_log


class PredefinedHandsLoader:
    """Loader for predefined hands configuration"""
    
    def __init__(self):
        self.config_path = os.path.join(
            os.path.dirname(__file__), 
            '..', 
            'config', 
            'predefined_hands.yaml'
        )
    
    def load_config(self) -> Dict[str, Any]:
        """
        Load the predefined hands configuration from YAML file
        
        Returns:
            Dict containing 'enabled' flag and 'hands' data
        """
        try:
            with open(self.config_path, 'r') as file:
                config = yaml.safe_load(file)
                custom_log(f"Loaded predefined hands config: enabled={config.get('enabled', False)}", 
                          level="INFO", isOn=True)
                return config
        except FileNotFoundError:
            custom_log("Predefined hands config file not found, using default (disabled)", 
                      level="WARNING", isOn=True)
            return {'enabled': False, 'hands': {}}
        except Exception as e:
            custom_log(f"Error loading predefined hands config: {e}", 
                      level="ERROR", isOn=True)
            return {'enabled': False, 'hands': {}}
    
    def get_hand_for_player(self, config: Dict[str, Any], player_index: int) -> Optional[List[Dict[str, str]]]:
        """
        Get predefined hand for a specific player
        
        Args:
            config: Configuration dictionary from load_config()
            player_index: Index of the player (0-based)
            
        Returns:
            List of card specifications or None if no predefined hand
        """
        if not config.get('enabled', False):
            return None
        
        hands = config.get('hands', {})
        if not hands:
            return None
        
        player_key = f"player_{player_index}"
        hand = hands.get(player_key)
        
        if hand is None:
            custom_log(f"No predefined hand found for player {player_index}", 
                      level="DEBUG", isOn=True)
            return None
        
        custom_log(f"Found predefined hand for player {player_index}: {len(hand)} cards", 
                  level="DEBUG", isOn=True)
        return hand

File: dart_bkend_base_01/lib/modules/recall_game/utils/platform/predefined_hands_loader.dart
import 'dart:io';
import 'package:yaml/yaml.dart';
import 'shared_imports.dart';

const bool LOGGING_SWITCH = false;


/// Predefined Hands Loader for Testing
/// 
/// This class provides functionality to load predefined hands configuration
/// for testing purposes in the Recall game.
class PredefinedHandsLoader {
  final Logger _logger = Logger();
  /// Load the predefined hands configuration from YAML file
  /// 
  /// Returns a Map containing 'enabled' flag and 'hands' data
  Future<Map<String, dynamic>> loadConfig() async {
    try {
      // Backend: attempt to read from a local file if present
      final file = File('lib/modules/recall_game/config/predefined_hands.yaml');
      if (!file.existsSync()) {
        return {'enabled': false, 'hands': {}};
      }
      final yamlString = await file.readAsString();
      final yamlMap = loadYaml(yamlString);
      final config = Map<String, dynamic>.from(yamlMap as Map);
      _logger.info('Loaded predefined hands config: enabled=${config['enabled']}', isOn: LOGGING_SWITCH);
      return config;
    } catch (e) {
      // Return disabled config if file doesn't exist
      _logger.warning('Predefined hands config load failed, defaulting to disabled: $e', isOn: LOGGING_SWITCH);
      return {'enabled': false, 'hands': {}};
    }
  }
  
  /// Get predefined hand for a specific player
  /// 
  /// [config] Configuration dictionary from loadConfig()
  /// [playerIndex] Index of the player (0-based)
  /// 
  /// Returns a List of card specifications or null if no predefined hand
  List<Map<String, String>>? getHandForPlayer(Map<String, dynamic> config, int playerIndex) {
    if (!config['enabled']) return null;
    
    final hands = config['hands'] as Map<dynamic, dynamic>?;
    if (hands == null) return null;
    
    final playerKey = 'player_$playerIndex';
    final hand = hands[playerKey] as List<dynamic>?;
    
    if (hand == null) {
      _logger.debug('No predefined hand found for player $playerIndex', isOn: LOGGING_SWITCH);
      return null;
    }
    
    _logger.debug('Found predefined hand for player $playerIndex: ${hand.length} cards', isOn: LOGGING_SWITCH);
    return hand.map((card) => {
      'rank': card['rank'].toString(),
      'suit': card['suit'].toString(),
    }).toList();
  }
}
